File: processing/dji_photogrammetry/drone_profiles.py
"""
Camera and flight profiles for supported DJI drones.

Use these presets to get accurate GSD / overlap calculations without
manually specifying sensor dimensions. Each profile also records the
hard limits the firmware enforces so the SDK can warn before a mission
upload is rejected.

Supported profiles
------------------
  mini3            – DJI Mini 3 (1/1.3" sensor, 12 MP)    — MSDK v5
  mini3_pro        – DJI Mini 3 Pro (1/1.3" sensor, 48 MP) — MSDK v5
  mini4_pro        – DJI Mini 4 Pro (1/1.3" sensor, 48 MP) — MSDK v5
  mavic3_classic   – DJI Mavic 3 Classic (4/3 sensor, 20 MP) — MSDK v5
  phantom4_rtk     – DJI Phantom 4 RTK (1" sensor, 20 MP, built-in RTK)
  generic          – Fallback for custom / unlisted cameras
"""

from dataclasses import dataclass, field
from typing import Optional, List


@dataclass
class FlightLimits:
    """Hard limits enforced by the drone firmware."""
    max_speed_ms: float
    min_altitude_m: float
    max_altitude_m: float
    max_waypoints: int
    has_obstacle_sensing: bool
    has_rtk: bool
    min_trigger_interval_s: float
    requires_msdk_v5: bool


@dataclass
class DroneProfile:
    """Complete camera + flight profile for one drone model."""
    name: str
    display_name: str
    focal_length_mm: float
    sensor_width_mm: float
    sensor_height_mm: float
    image_width_px: int
    image_height_px: int
    gimbal_pitch_deg: float
    limits: FlightLimits


@dataclass
class ValidationResult:
    """Result of validating MissionSettings against a DroneProfile."""
    is_valid: bool
    errors: List[str]
    warnings: List[str]


# ---------------------------------------------------------------------------
# Profiles
# ---------------------------------------------------------------------------

PROFILES = {

    # -----------------------------------------------------------------------
    # DJI Mini 3
    # -----------------------------------------------------------------------
    # Sensor: 1/1.3" CMOS  |  9.6 × 7.2 mm
    # Physical focal length: 6.7 mm  (24 mm 35mm-equivalent)
    # Resolution: 12 MP  (4032 × 3024)
    # Max speed: 16 m/s sport; 10 m/s normal (use normal for photo missions)
    # Weight: 249 g — under most country registration thresholds
    # MSDK: v5 only  |  No built-in RTK
    # Obstacle sensing: forward + rear + downward (NOT side)
    # -----------------------------------------------------------------------
    "mini3": DroneProfile(
        name="mini3",
        display_name="DJI Mini 3",
        focal_length_mm=6.7,
        sensor_width_mm=9.6,
        sensor_height_mm=7.2,
        image_width_px=4032,
        image_height_px=3024,
        gimbal_pitch_deg=-90.0,
        limits=FlightLimits(
            max_speed_ms=10.0,
            min_altitude_m=2.0,
            max_altitude_m=120.0,
            max_waypoints=65535,
            has_obstacle_sensing=True,
            has_rtk=False,
            min_trigger_interval_s=0.7,
            requires_msdk_v5=True,
        ),
    ),

    # -----------------------------------------------------------------------
    # DJI Mini 3 Pro
    # -----------------------------------------------------------------------
    # Sensor: 1/1.3" CMOS  |  9.6 × 7.2 mm
    # Resolution: 48 MP  (8064 × 6048)  or  12 MP binned
    # Obstacle sensing: omnidirectional (trinocular front/rear + side)
    # MSDK: v5 only  |  No built-in RTK
    # -----------------------------------------------------------------------
    "mini3_pro": DroneProfile(
        name="mini3_pro",
        display_name="DJI Mini 3 Pro",
        focal_length_mm=6.7,
        sensor_width_mm=9.6,
        sensor_height_mm=7.2,
        image_width_px=8064,
        image_height_px=6048,
        gimbal_pitch_deg=-90.0,
        limits=FlightLimits(
            max_speed_ms=10.0,
            min_altitude_m=2.0,
            max_altitude_m=120.0,
            max_waypoints=65535,
            has_obstacle_sensing=True,
            has_rtk=False,
            min_trigger_interval_s=0.7,
            requires_msdk_v5=True,
        ),
    ),

    # -----------------------------------------------------------------------
    # DJI Mini 4 Pro
    # -----------------------------------------------------------------------
    # Sensor: 1/1.3" CMOS  |  9.6 × 7.2 mm
    # Resolution: 48 MP  (8064 × 6048)
    # Obstacle sensing: omnidirectional
    # MSDK: v5 only  |  No built-in RTK
    # -----------------------------------------------------------------------
    "mini4_pro": DroneProfile(
        name="mini4_pro",
        display_name="DJI Mini 4 Pro",
        focal_length_mm=6.7,
        sensor_width_mm=9.6,
        sensor_height_mm=7.2,
        image_width_px=8064,
        image_height_px=6048,
        gimbal_pitch_deg=-90.0,
        limits=FlightLimits(
            max_speed_ms=10.0,
            min_altitude_m=2.0,
            max_altitude_m=120.0,
            max_waypoints=65535,
            has_obstacle_sensing=True,
            has_rtk=False,
            min_trigger_interval_s=0.7,
            requires_msdk_v5=True,
        ),
    ),

    # -----------------------------------------------------------------------
    # DJI Mavic 3 Classic
    # -----------------------------------------------------------------------
    # Sensor: 4/3 CMOS (Hasselblad)  |  17.3 × 13.0 mm
    # Physical focal length: 12.29 mm  (24 mm equivalent)
    # Resolution: 20 MP  (5280 × 3956)
    # MSDK: v5  |  No built-in RTK
    # -----------------------------------------------------------------------
    "mavic3_classic": DroneProfile(
        name="mavic3_classic",
        display_name="DJI Mavic 3 Classic",
        focal_length_mm=12.29,
        sensor_width_mm=17.3,
        sensor_height_mm=13.0,
        image_width_px=5280,
        image_height_px=3956,
        gimbal_pitch_deg=-90.0,
        limits=FlightLimits(
            max_speed_ms=15.0,
            min_altitude_m=2.0,
            max_altitude_m=120.0,
            max_waypoints=65535,
            has_obstacle_sensing=True,
            has_rtk=False,
            min_trigger_interval_s=0.5,
            requires_msdk_v5=True,
        ),
    ),

    # -----------------------------------------------------------------------
    # DJI Mavic 3 Multispectral
    # -----------------------------------------------------------------------
    # Sensor: 4/3 CMOS (same optics as Mavic 3 Classic) + 4-band MS camera
    # MS sensor: 1/2.8" per band  |  2.08 MP each
    # Use the RGB camera settings for overlap calculations
    # MSDK: v5  |  Optional RTK via D-RTK 2
    # -----------------------------------------------------------------------
    "mavic3_multispectral": DroneProfile(
        name="mavic3_multispectral",
        display_name="DJI Mavic 3 Multispectral",
        focal_length_mm=12.29,
        sensor_width_mm=17.3,
        sensor_height_mm=13.0,
        image_width_px=5280,
        image_height_px=3956,
        gimbal_pitch_deg=-90.0,
        limits=FlightLimits(
            max_speed_ms=15.0,
            min_altitude_m=2.0,
            max_altitude_m=120.0,
            max_waypoints=65535,
            has_obstacle_sensing=True,
            has_rtk=True,
            min_trigger_interval_s=0.7,
            requires_msdk_v5=True,
        ),
    ),

    # -----------------------------------------------------------------------
    # DJI Phantom 4 RTK
    # -----------------------------------------------------------------------
    # Sensor: 1" CMOS  |  13.2 × 8.8 mm
    # Physical focal length: 8.8 mm  (24 mm equivalent)
    # Resolution: 20 MP  (5472 × 3648)
    # Built-in RTK (D-RTK 2 base or NTRIP)
    # MSDK: v4 and v5 compatible
    # -----------------------------------------------------------------------
    "phantom4_rtk": DroneProfile(
        name="phantom4_rtk",
        display_name="DJI Phantom 4 RTK",
        focal_length_mm=8.8,
        sensor_width_mm=13.2,
        sensor_height_mm=8.8,
        image_width_px=5472,
        image_height_px=3648,
        gimbal_pitch_deg=-90.0,
        limits=FlightLimits(
            max_speed_ms=15.0,
            min_altitude_m=2.0,
            max_altitude_m=120.0,
            max_waypoints=65535,
            has_obstacle_sensing=False,
            has_rtk=True,
            min_trigger_interval_s=0.5,
            requires_msdk_v5=False,
        ),
    ),

    # -----------------------------------------------------------------------
    # Generic / custom camera
    # -----------------------------------------------------------------------
    "generic": DroneProfile(
        name="generic",
        display_name="Generic / Custom",
        focal_length_mm=24.0,
        sensor_width_mm=13.2,
        sensor_height_mm=8.8,
        image_width_px=4000,
        image_height_px=3000,
        gimbal_pitch_deg=-90.0,
        limits=FlightLimits(
            max_speed_ms=15.0,
            min_altitude_m=2.0,
            max_altitude_m=120.0,
            max_waypoints=65535,
            has_obstacle_sensing=False,
            has_rtk=False,
            min_trigger_interval_s=0.5,
            requires_msdk_v5=False,
        ),
    ),
}


def get_profile(name: str) -> DroneProfile:
    """
    Return a drone profile by name (case-insensitive).

    Args:
        name: Profile key, e.g. "mini3", "mini3_pro", "phantom4_rtk".

    Returns:
        DroneProfile dataclass.

    Raises:
        ValueError: If name is not recognised. Lists available names.
    """
    key = name.lower().replace("-", "_").replace(" ", "_")
    if key not in PROFILES:
        available = ", ".join(sorted(PROFILES.keys()))
        raise ValueError(
            f"Unknown drone profile '{name}'. "
            f"Available profiles: {available}"
        )
    return PROFILES[key]


def list_profiles() -> List[str]:
    """Return sorted list of available profile names."""
    return sorted(PROFILES.keys())


def validate_mission(profile: DroneProfile, altitude_m: float,
                     speed_ms: float,
                     trigger_interval_s: Optional[float] = None) -> ValidationResult:
    """
    Validate proposed mission parameters against a drone profile's limits.

    Returns a ValidationResult with .is_valid, .errors, and .warnings.
    Always call this before generating waypoints and surface any errors
    to the user — the firmware will silently reject out-of-range values.
    """
    errors: List[str]   = []
    warnings: List[str] = []
    lim = profile.limits

    # Speed
    if speed_ms > lim.max_speed_ms:
        errors.append(
            f"Speed {speed_ms} m/s exceeds {profile.display_name} "
            f"maximum of {lim.max_speed_ms} m/s. "
            f"Reduce to {lim.max_speed_ms} m/s or less."
        )

    # Altitude
    if altitude_m < lim.min_altitude_m:
        errors.append(
            f"Altitude {altitude_m} m is below the {profile.display_name} "
            f"minimum of {lim.min_altitude_m} m AGL."
        )
    if altitude_m > lim.max_altitude_m:
        warnings.append(
            f"Altitude {altitude_m} m exceeds the recommended "
            f"{lim.max_altitude_m} m AGL regulatory limit in most regions."
        )

    # Trigger interval
    if trigger_interval_s is not None and trigger_interval_s < lim.min_trigger_interval_s:
        errors.append(
            f"Trigger interval {trigger_interval_s} s is below the "
            f"{profile.display_name} minimum of {lim.min_trigger_interval_s} s."
        )

    # Consumer Mini series — specific advisories
    if profile.name in ("mini3", "mini3_pro", "mini4_pro"):
        if not lim.has_rtk:
            warnings.append(
                f"{profile.display_name} has no built-in RTK. For survey-grade "
                f"accuracy use Ground Control Points (GCPs) or a PPK workflow."
            )
        if lim.requires_msdk_v5:
            warnings.append(
                f"{profile.display_name} requires DJI Mobile SDK v5. "
                f"MSDK v4 is not compatible with this aircraft."
            )
        warnings.append(
            f"{profile.display_name} weighs 249 g. Even sub-250 g drones are "
            f"subject to 120 m AGL limits in most countries — check local rules."
        )

    return ValidationResult(is_valid=len(errors) == 0, errors=errors, warnings=warnings)

File: mobile/android-sdk-wrapper/src/main/java/com/dji/photogrammetry/DroneProfiles.kt
package com.dji.photogrammetry

/**
 * Camera and flight profiles for supported DJI drones.
 *
 * Each profile captures the real-world sensor geometry needed for accurate
 * GSD / overlap calculations, plus the hard limits the firmware enforces
 * so the SDK can warn or clamp values before uploading a mission.
 *
 * Supported profiles
 * ------------------
 *  MINI_3           – DJI Mini 3 (1/1.3" sensor, 24 mm equiv, 12 MP)
 *  MINI_3_PRO       – DJI Mini 3 Pro (1/1.3" sensor, 24 mm equiv, 48 MP)
 *  MINI_4_PRO       – DJI Mini 4 Pro (1/1.3" sensor, 24 mm equiv, 48 MP)
 *  MAVIC_3_CLASSIC  – DJI Mavic 3 Classic (4/3 sensor, 24 mm equiv, 20 MP)
 *  MAVIC_3_MULTISPECTRAL – Mavic 3 Multispectral (4/3 + 4-band MS sensor)
 *  PHANTOM_4_RTK    – DJI Phantom 4 RTK (1" sensor, 24 mm equiv, 20 MP)
 *  GENERIC          – Fallback for unlisted cameras
 */

enum class DroneModel {
    MINI_3,
    MINI_3_PRO,
    MINI_4_PRO,
    MAVIC_3_CLASSIC,
    MAVIC_3_MULTISPECTRAL,
    PHANTOM_4_RTK,
    GENERIC
}

data class FlightLimits(
    /** Maximum mission flight speed (m/s). Firmware will reject higher values. */
    val maxSpeedMs: Double,
    /** Minimum safe AGL altitude (m). */
    val minAltitudeM: Double,
    /** Maximum AGL altitude before regulatory warning (m). */
    val maxAltitudeM: Double,
    /** Maximum waypoints per WaypointV2 mission. */
    val maxWaypoints: Int,
    /** True when the drone has obstacle-sensing (forward + rear at minimum). */
    val hasObstacleSensing: Boolean,
    /** True when built-in RTK receiver is present. */
    val hasRtk: Boolean,
    /** Minimum camera trigger interval (seconds). */
    val minTriggerIntervalS: Double,
    /**
     * True when this model requires MSDK v5.
     * Mini 3, Mini 3 Pro, Mini 4 Pro, Mavic 3 series all require MSDK v5.
     * Phantom 4 RTK works with both v4 and v5.
     */
    val requiresMsdkV5: Boolean
)

data class DroneProfile(
    val model: DroneModel,
    val displayName: String,
    val cameraSettings: CameraSettings,
    val limits: FlightLimits
)

object DroneProfiles {

    // -----------------------------------------------------------------------
    // DJI Mini 3
    // -----------------------------------------------------------------------
    //  Sensor: 1/1.3" CMOS (9.6 mm × 7.2 mm)
    //  Physical focal length: 6.7 mm
    //  35 mm equivalent: 24 mm
    //  Effective pixels: 12 MP  (4032 × 3024)
    //  Max photo resolution: 4032 × 3024
    //  Video: 4K/60fps
    //  Max speed: 16 m/s sport; 10 m/s normal; 6 m/s cine
    //  Max flight time: 38 min
    //  Weight: 249 g (under most country registration thresholds)
    //  MSDK: v5 only
    //  Notes: NO obstacle sensing on sides; forward + downward + rear only
    // -----------------------------------------------------------------------
    val MINI_3 = DroneProfile(
        model = DroneModel.MINI_3,
        displayName = "DJI Mini 3",
        cameraSettings = CameraSettings(
            focalLengthMm   = 6.7,
            sensorWidthMm   = 9.6,
            sensorHeightMm  = 7.2,
            imageWidthPx    = 4032,
            imageHeightPx   = 3024,
            gimbalPitchDeg  = -90.0
        ),
        limits = FlightLimits(
            maxSpeedMs           = 10.0,   // 10 m/s in Normal mode (recommended for photo)
            minAltitudeM         = 2.0,
            maxAltitudeM         = 120.0,  // 120 m AGL — regulatory default in most regions
            maxWaypoints         = 65535,  // MSDK v5 WaypointV2 supports up to 65535
            hasObstacleSensing   = true,
            hasRtk               = false,
            minTriggerIntervalS  = 0.7,    // Minimum shutter interval at this resolution
            requiresMsdkV5       = true
        )
    )

    // -----------------------------------------------------------------------
    // DJI Mini 3 Pro
    // -----------------------------------------------------------------------
    //  Sensor: 1/1.3" CMOS (9.6 mm × 7.2 mm)
    //  Physical focal length: 6.7 mm
    //  Effective pixels: 48 MP  (8064 × 6048) or 12 MP quad-bayer binned
    //  Obstacle sensing: omnidirectional (trinocular front/rear + side)
    //  MSDK: v5 only
    // -----------------------------------------------------------------------
    val MINI_3_PRO = DroneProfile(
        model = DroneModel.MINI_3_PRO,
        displayName = "DJI Mini 3 Pro",
        cameraSettings = CameraSettings(
            focalLengthMm   = 6.7,
            sensorWidthMm   = 9.6,
            sensorHeightMm  = 7.2,
            imageWidthPx    = 8064,    // 48 MP native
            imageHeightPx   = 6048,
            gimbalPitchDeg  = -90.0
        ),
        limits = FlightLimits(
            maxSpeedMs           = 10.0,
            minAltitudeM         = 2.0,
            maxAltitudeM         = 120.0,
            maxWaypoints         = 65535,
            hasObstacleSensing   = true,
            hasRtk               = false,
            minTriggerIntervalS  = 0.7,
            requiresMsdkV5       = true
        )
    )

    // -----------------------------------------------------------------------
    // DJI Mini 4 Pro
    // -----------------------------------------------------------------------
    //  Sensor: 1/1.3" CMOS (9.6 mm × 7.2 mm)
    //  Physical focal length: 6.7 mm (main); also has 70 mm tele
    //  Effective pixels: 48 MP  (8064 × 6048)
    //  Obstacle sensing: omnidirectional
    //  MSDK: v5 only
    // -----------------------------------------------------------------------
    val MINI_4_PRO = DroneProfile(
        model = DroneModel.MINI_4_PRO,
        displayName = "DJI Mini 4 Pro",
        cameraSettings = CameraSettings(
            focalLengthMm   = 6.7,
            sensorWidthMm   = 9.6,
            sensorHeightMm  = 7.2,
            imageWidthPx    = 8064,
            imageHeightPx   = 6048,
            gimbalPitchDeg  = -90.0
        ),
        limits = FlightLimits(
            maxSpeedMs           = 10.0,
            minAltitudeM         = 2.0,
            maxAltitudeM         = 120.0,
            maxWaypoints         = 65535,
            hasObstacleSensing   = true,
            hasRtk               = false,
            minTriggerIntervalS  = 0.7,
            requiresMsdkV5       = true
        )
    )

    // -----------------------------------------------------------------------
    // DJI Mavic 3 Classic
    // -----------------------------------------------------------------------
    //  Sensor: 4/3 CMOS (17.3 mm × 13 mm) — Hasselblad L-Format
    //  Physical focal length: 12.29 mm / 35 mm equiv: 24 mm
    //  Effective pixels: 20 MP  (5280 × 3956)
    //  MSDK: v5
    // -----------------------------------------------------------------------
    val MAVIC_3_CLASSIC = DroneProfile(
        model = DroneModel.MAVIC_3_CLASSIC,
        displayName = "DJI Mavic 3 Classic",
        cameraSettings = CameraSettings(
            focalLengthMm   = 12.29,
            sensorWidthMm   = 17.3,
            sensorHeightMm  = 13.0,
            imageWidthPx    = 5280,
            imageHeightPx   = 3956,
            gimbalPitchDeg  = -90.0
        ),
        limits = FlightLimits(
            maxSpeedMs           = 15.0,
            minAltitudeM         = 2.0,
            maxAltitudeM         = 120.0,
            maxWaypoints         = 65535,
            hasObstacleSensing   = true,
            hasRtk               = false,
            minTriggerIntervalS  = 0.5,
            requiresMsdkV5       = true
        )
    )

    // -----------------------------------------------------------------------
    // DJI Phantom 4 RTK
    // -----------------------------------------------------------------------
    //  Sensor: 1" CMOS (13.2 mm × 8.8 mm)
    //  Physical focal length: 8.8 mm / 35 mm equiv: 24 mm
    //  Effective pixels: 20 MP  (5472 × 3648)
    //  Built-in RTK receiver (D-RTK 2 base or NTRIP)
    //  MSDK: v4 and v5 compatible
    // -----------------------------------------------------------------------
    val PHANTOM_4_RTK = DroneProfile(
        model = DroneModel.PHANTOM_4_RTK,
        displayName = "DJI Phantom 4 RTK",
        cameraSettings = CameraSettings(
            focalLengthMm   = 8.8,
            sensorWidthMm   = 13.2,
            sensorHeightMm  = 8.8,
            imageWidthPx    = 5472,
            imageHeightPx   = 3648,
            gimbalPitchDeg  = -90.0
        ),
        limits = FlightLimits(
            maxSpeedMs           = 15.0,
            minAltitudeM         = 2.0,
            maxAltitudeM         = 120.0,
            maxWaypoints         = 65535,
            hasObstacleSensing   = false,
            hasRtk               = true,
            minTriggerIntervalS  = 0.5,
            requiresMsdkV5       = false
        )
    )

    // -----------------------------------------------------------------------
    // Generic / custom camera
    // -----------------------------------------------------------------------
    val GENERIC = DroneProfile(
        model = DroneModel.GENERIC,
        displayName = "Generic / Custom",
        cameraSettings = CameraSettings(
            focalLengthMm   = 24.0,
            sensorWidthMm   = 13.2,
            sensorHeightMm  = 8.8,
            imageWidthPx    = 4000,
            imageHeightPx   = 3000,
            gimbalPitchDeg  = -90.0
        ),
        limits = FlightLimits(
            maxSpeedMs           = 15.0,
            minAltitudeM         = 2.0,
            maxAltitudeM         = 120.0,
            maxWaypoints         = 65535,
            hasObstacleSensing   = false,
            hasRtk               = false,
            minTriggerIntervalS  = 0.5,
            requiresMsdkV5       = false
        )
    )

    /** Look up a profile by model enum. */
    fun forModel(model: DroneModel): DroneProfile = when (model) {
        DroneModel.MINI_3                -> MINI_3
        DroneModel.MINI_3_PRO            -> MINI_3_PRO
        DroneModel.MINI_4_PRO            -> MINI_4_PRO
        DroneModel.MAVIC_3_CLASSIC       -> MAVIC_3_CLASSIC
        DroneModel.MAVIC_3_MULTISPECTRAL -> MAVIC_3_CLASSIC   // same optics
        DroneModel.PHANTOM_4_RTK         -> PHANTOM_4_RTK
        DroneModel.GENERIC               -> GENERIC
    }

    /** Returns a validation result for proposed mission settings on this drone. */
    fun validate(profile: DroneProfile, settings: MissionSettings): ValidationResult {
        val warnings  = mutableListOf<String>()
        val errors    = mutableListOf<String>()
        val clamped   = settings.copy()

        // Speed check
        if (settings.flightSpeedMs > profile.limits.maxSpeedMs) {
            errors.add("Speed ${settings.flightSpeedMs} m/s exceeds ${profile.displayName} " +
                    "max of ${profile.limits.maxSpeedMs} m/s. " +
                    "Reduce to ${profile.limits.maxSpeedMs} m/s or less.")
        }

        // Altitude checks
        if (settings.altitudeM < profile.limits.minAltitudeM) {
            errors.add("Altitude ${settings.altitudeM} m is below the minimum " +
                    "${profile.limits.minAltitudeM} m AGL.")
        }
        if (settings.altitudeM > profile.limits.maxAltitudeM) {
            warnings.add("Altitude ${settings.altitudeM} m exceeds the recommended " +
                    "${profile.limits.maxAltitudeM} m AGL regulatory limit in most regions.")
        }

        // Mini 3 specific
        if (profile.model == DroneModel.MINI_3 || profile.model == DroneModel.MINI_3_PRO ||
            profile.model == DroneModel.MINI_4_PRO
        ) {
            if (settings.altitudeM > 120.0) {
                warnings.add("${profile.displayName} at ${settings.altitudeM} m: in most " +
                        "countries 249 g drones are still subject to 120 m AGL limits. " +
                        "Check local regulations.")
            }
            if (!profile.limits.hasRtk) {
                warnings.add("${profile.displayName} has no built-in RTK. For survey-grade " +
                        "accuracy use GCPs or an external PPK workflow.")
            }
            if (profile.limits.requiresMsdkV5) {
                warnings.add("${profile.displayName} requires DJI Mobile SDK v5. " +
                        "MSDK v4 is not compatible with this aircraft.")
            }
        }

        // Trigger interval
        settings.triggerTimeS?.let { ts ->
            if (ts < profile.limits.minTriggerIntervalS) {
                errors.add("Trigger interval ${ts} s is below the minimum " +
                        "${profile.limits.minTriggerIntervalS} s for ${profile.displayName}.")
            }
        }

        return ValidationResult(
            isValid   = errors.isEmpty(),
            errors    = errors,
            warnings  = warnings,
            clampedSettings = clamped
        )
    }
}

data class ValidationResult(
    val isValid: Boolean,
    val errors: List<String>,
    val warnings: List<String>,
    val clampedSettings: MissionSettings
)
